src/simulators/unitary/unitarymatrix.hpp

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2018, IBM.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 found in
+ * the LICENSE.txt file in the root directory of this source tree.
+ */
+
+#ifndef _qv_unitary_matrix_hpp_
+#define _qv_unitary_matrix_hpp_
+
+
+#include "framework/utils.hpp"
+#include "simulators/statevector/qubitvector.hpp"
+
+namespace QV {
+
+//============================================================================
+// UnitaryMatrix class
+//============================================================================
+
+// This class is derived from the QubitVector class and stores an N-qubit 
+// matrix as a 2*N-qubit vector.
+// The vector is formed using column-stacking vectorization as under this
+// convention left-matrix multiplication on qubit-n is equal to multiplication
+// of the vectorized 2*N qubit vector also on qubit-n.
+
+template <class data_t = complex_t*>
+class UnitaryMatrix : public QubitVector<data_t> {
+
+public:
+  // Type aliases
+  using BaseVector = QubitVector<data_t>;
+
+  //-----------------------------------------------------------------------
+  // Constructors and Destructor
+  //-----------------------------------------------------------------------
+
+  UnitaryMatrix() : UnitaryMatrix(0) {};
+  explicit UnitaryMatrix(size_t num_qubits);
+  UnitaryMatrix(const UnitaryMatrix& obj) = delete;
+  UnitaryMatrix &operator=(const UnitaryMatrix& obj) = delete;
+
+  //-----------------------------------------------------------------------
+  // Utility functions
+  //-----------------------------------------------------------------------
+
+  // Set the size of the vector in terms of qubit number
+  inline void set_num_qubits(size_t num_qubits);
+
+  // Return the number of rows in the matrix
+  size_t num_rows() const {return rows_;}
+
+  // Returns the number of qubits for the current vector
+  inline uint_t num_qubits() const { return num_qubits_;}
+
+  // Returns a copy of the underlying data_t data as a complex vector
+  AER::cmatrix_t matrix() const;
+
+  // Return JSON serialization of UnitaryMatrix;
+  json_t json() const;
```

Comment:
At some point in the future, we have to review all the responsibilities our classes have because they might be too many.
In general, enforcing the `Single Responsibility Principle` is quite beneficial for having a flexible, debuggable, hence maintainable code base.

Reply:
In this case for example, the `UnitaryMatrix` should know nothing about `json` or serialization... this is other's component responsibility, that will probably know how to serialize any of our components, not only a `UnitaryMatrix`.